src/pptrees.cc

```diff
@@ -1,26 +1,279 @@
+	std::string adder_name = tempdir_name.substr(19);
+
+	// Define python executable
+	std::string python_executable = "python3";
+
+	// Define adder generation script locally, to avoid dealing with an extra file
+	// Import modules
+	std::string py_command = "import sys; from pptrees.yosys_alu import yosys_alu as tree; ";
+	// Initialize the tree
+	py_command.append(stringf("g=tree(%d,'%s'); ", width, start.c_str()));
+	// Apply each transform
+	py_command.append("[getattr(g,t.split('@')[0])(*[int(x) for x in t.split('@')[1].split(',')]) ");
+	// Loop over the transforms
+	py_command.append(stringf("for t in '%s'.split('_')[1:]];", transform.c_str()));
+	// Output the HDL
+	py_command.append(stringf("g.hdl('%s/pptrees_alu.v','%s',top_module='_%s_adder');",tempdir_name.c_str(),mapping.c_str(),adder_name.c_str()));
+
+	// Call python
+	std::string shell_command = stringf("%s -c \"%s\"", python_executable.c_str(), py_command.c_str());
+
+	log("Running python script from shell: %s\n", shell_command.c_str());
+	int ret = run_command(shell_command, std::bind(&python_output_filter::next_line, filt, std::placeholders::_1));
+
+	if (ret != 0)
+		log_error("Synth_opt_adder: execution of command \"%s\" failed: return code %d.\n", py_command.c_str(), ret);
+
+	return tempdir_name;
+}
+
+struct opt_pptrees : public Pass {
+	opt_pptrees() : Pass("opt_pptrees", "Optimize parallel prefix trees using pptrees Python package") { }
+	
+	void help() override
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("	opt_pptrees [options] [selection]\n");
+		log("\n");
+		log("This pass uses the synth_opt_adders tool to optimize adders\n");
+		log("\n");
+		log("	-mapping <string>\n");
+		log("		mapping strategy (default: behavioral).\n");
+		log("\n");
+
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design *design) override
+	{
+		log_header(design, "Executing opt_pptrees pass.\n");
+		log_push();
+
+        int width;
+        std::string start = "ripple-carry";
+        std::string transforms = "";
+        std::string mapping = "behavioral";
+
+	// Parse options
+        size_t argidx;
+        for (argidx = 1; argidx < args.size(); argidx++) {
+            std::string arg = args[argidx];
+		if (arg == "-mapping" && argidx+1 < args.size()) {
+			mapping = args[++argidx]; //super fragile; do this better
+		}
         }
+        
+        extra_args(args, argidx, design);
+
+	// Write out mappping files
+	std::string mapping_dir = python_maps(mapping);
+
+	// Get the output of the python script
+	Pass::call(design, stringf("read_verilog -sv %s/%s_map.v", mapping_dir.c_str(), mapping.c_str()));
+	Pass::call(design, stringf("read_verilog -sv %s/modules.v", mapping_dir.c_str()));
+
+	// Save attributes of $add cells before they are removed by alumacc
+	dict<std::string, dict<RTLIL::IdString, RTLIL::Const>> saved_attributes = save_attributes(design);
```

Comment:
This saves attributes for the entire design, even if only a portion is selected.  That may be fine, my intuition is that the overhead is small and I can't think of anything it would break, but it's good to keep in mind.